### PR TITLE
feat(osio): Introduce new `/osio/launch/empty` API end-point

### PR DESCRIPTION
**Pull request by ravsa**
A new POST API endpoint `/osio/launch/empty` is added which take payload same as `/osio/launch`  endpoint and creates a repository with no content.

### CODE REVIEW

**Comment by alien-ike** [comment]
### Ike Plugins (test-keeper)

Thank you @ravsa for this contribution!



It appears that no tests have been added or updated in this PR.

Automated tests give us confidence in shipping reliable software. Please add some as part of this change.

If you are an admin or the reviewer of this PR and you are sure that no test is needed then you can use the command `/ok-without-tests` as a comment to make the status green.


For more information please head over to official [documentation](http://arquillian.org/ike-prow-plugins/#_test_keeper_plugin). You can find there how to configure the plugin - for example exclude certain file types so if PR contains only them it won't be checked.

File: addons/osio-addon/src/main/java/io/fabric8/launcher/osio/OsioMissionControl.java [hunk 1]
```diff
@@ -166,4 +166,39 @@ public Boom launchImport(OsioImportProjectile projectile) {
                 .createdProject(ImmutableOpenShiftProject.builder().name(projectile.getOpenShiftProjectName()).build())
                 .build();
     }
+
+    /**
+     * Used in /osio/launch/empty
+     */
+    public Boom launchEmpty(Projectile genericProjectile) throws IllegalArgumentException {
```
Comment (gastaldi) [inline addons/osio-addon/src/main/java/io/fabric8/launcher/osio/OsioMissionControl.java]: Change the parameter type to OsioLaunchProjectile

**Comment by gastaldi** [comment]
Also, don't forget to update the swagger.yaml documenting this new endpoint

File: addons/osio-addon/src/main/java/io/fabric8/launcher/osio/OsioEmptyLaunchMissionControl.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+import io.fabric8.launcher.osio.steps.GitSteps;
+import io.fabric8.launcher.osio.steps.OpenShiftSteps;
+import io.fabric8.launcher.osio.steps.WitSteps;
+import io.fabric8.launcher.service.git.api.GitRepository;
+import io.fabric8.launcher.service.openshift.api.ImmutableOpenShiftProject;
+import io.fabric8.openshift.api.model.BuildConfig;
+
+
+@Dependent
+public class OsioEmptyLaunchMissionControl implements MissionControl<OsioProjectileContext, OsioLaunchProjectile> {
```
Comment (gastaldi) [inline addons/osio-addon/src/main/java/io/fabric8/launcher/osio/OsioEmptyLaunchMissionControl.java]: How is this different from `OsioLaunchMissionControl`?

File: web/src/main/webapp/swagger.yaml [hunk 3]
```diff
@@ -770,6 +770,104 @@ paths:
...
+                type: array
+                items:
+                  $ref: '#/components/schemas/ValidationError'
+      requestBody:
+        content:
+          application/x-www-form-urlencoded:
+            schema:
+              type: object
+              properties:
+                mission:
```
Comment (gastaldi) [inline web/src/main/webapp/swagger.yaml]: Are you still requiring mission in this case?

File: addons/osio-addon/src/main/java/io/fabric8/launcher/osio/OsioEmptyLaunchMissionControl.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+                .pipelineId(context.getPipelineId())
+                .build();
+    }
+
+    @Override
+    public Boom launch(OsioLaunchProjectile projectile) {
+        // Make sure that cd-github is created in Openshift
+        openShiftSteps.ensureCDGithubSecretExists();
+
+        final GitRepository repository = gitSteps.createRepository(projectile);
```
Comment (gastaldi) [inline addons/osio-addon/src/main/java/io/fabric8/launcher/osio/OsioEmptyLaunchMissionControl.java]: Nothing is being pushed to the repository, is that intended?

Comment (ravsa) [reply addons/osio-addon/src/main/java/io/fabric8/launcher/osio/OsioEmptyLaunchMissionControl.java]: yes, we need repository with no content.

File: web/src/main/webapp/swagger.yaml (hunk 3 above)
Comment (ravsa) [reply web/src/main/webapp/swagger.yaml]: yes, we require existing mission like `health-check` or `rest-http` as I discussed with you earlier.

File: addons/osio-addon/src/main/java/io/fabric8/launcher/osio/OsioEmptyLaunchMissionControl.java (hunk 2 above)
Comment (ravsa) [reply addons/osio-addon/src/main/java/io/fabric8/launcher/osio/OsioEmptyLaunchMissionControl.java]: The only difference is that we are not pushing anything to Github repo.

File: addons/osio-addon/src/main/java/io/fabric8/launcher/osio/OsioEmptyLaunchMissionControl.java (hunk 4 above)
Comment (gastaldi) [reply addons/osio-addon/src/main/java/io/fabric8/launcher/osio/OsioEmptyLaunchMissionControl.java]: In this case why bother calling prepare and choosing a specific mission? This doesn't make much sense to me.

File: web/src/main/webapp/swagger.yaml [hunk 5]
```diff
@@ -767,6 +767,9 @@ paths:
                 projectName:
                   description: The name of the project to create
                   type: string
+                emptyRepository:
```
Comment (gastaldi) [inline web/src/main/webapp/swagger.yaml]: `emptyGitRepository` seems more readable

File: web/src/test/java/io/fabric8/launcher/web/endpoints/osio/OsioEmptyLaunchEndpointIT.java [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+
+import static io.fabric8.launcher.web.endpoints.osio.OsioTests.getOsioIdentity;
+import static io.fabric8.launcher.web.endpoints.osio.OsioTests.getWitClient;
+import static io.restassured.RestAssured.given;
+import static org.apache.commons.lang3.RandomStringUtils.randomAlphanumeric;
+import static org.assertj.core.api.Assertions.assertThat;
+
+@RunWith(Arquillian.class)
+@RunAsClient
+public class OsioEmptyLaunchEndpointIT {
```
Comment (gastaldi) [inline web/src/test/java/io/fabric8/launcher/web/endpoints/osio/OsioEmptyLaunchEndpointIT.java]: How is this test asserting that the repository created is empty?

**Comment by gastaldi** [comment]
There are some test failures. See the logs in https://semaphoreci.com/fabric8-launcher/launcher-backend/branches/pull-request-431/builds/10

File: web/src/test/java/io/fabric8/launcher/web/endpoints/osio/OsioEmptyLaunchEndpointIT.java [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+
+    private static Space space;
+
+    @Deployment(testable = false)
+    public static Archive<?> createDeployment() {
+        return ShrinkWrap.createFromZipFile(WebArchive.class, new File("target/launcher-backend.war"));
+    }
+
+    @Test
+    public void should_launch() throws Exception {
```
Comment (gastaldi) [inline web/src/test/java/io/fabric8/launcher/web/endpoints/osio/OsioEmptyLaunchEndpointIT.java]: Move this test to OsioLaunchEndpointIT under a new method and make sure it tests that the repository is empty

**Comment by gastaldi** [comment]
Thanks!

**Comment by quintesse** [comment]
What is this used for btw? It seems strange to be creating an empty git repository?

**Pull request merged**